### PR TITLE
Feature/DLLscripting

### PR DESCRIPTION
**Pull request by AlbertVVila**
This is a first version of the scripting feature, to test correctly this pull request, follow these steps:

1. Update VS to last version.
2. Pull this branch.
4. Compile the Engine.
5. Compile the Player Movement Script and test it using the play button and moving (WASD). 😄 
6. To create your custom script, first include [TemplateScript.zip](https://github.com/fractal-puppy/Engine/files/2900447/TemplateScript.zip) inside Documents/Visual Studio 2017/Templates/ProjectTemplates.
7. Then follow this [Tutorial](https://drive.google.com/open?id=1zUJc-IFYMmCiGMncUAb54r1TrgaChWkG).
8. 😃

### CODE REVIEW

**Comment by AlbertVVila** [comment]
Imgui.ini was included in order to setup the play & pause layout

File: Scripts/BaseScript/BaseScript.h [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+#ifdef BASESCRIPT_EXPORTS
+#define  BASESCRIPT_API __declspec(dllexport)
+#else
+#define BASESCRIPT_API __declspec(dllimport)
+#endif
+
+class BASESCRIPT_API Script
+{
+public:
+	void SetApp(Application * app);
```
Comment (sondy99) [inline Scripts/BaseScript/BaseScript.h]: CodeStyle: * close to the type

File: Scripts/BaseScript/BaseScript.h [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+	void SetGameObject(GameObject* go);
+	virtual void Expose(ImGuiContext *context);
```
Comment (sondy99) [inline Scripts/BaseScript/BaseScript.h]: CodeStyle: * close to the type

File: Scripts/BaseScript/BaseScript.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+	virtual void Serialize(JSON_value* json) const {}
+	virtual void DeSerialize(JSON_value * json) {}
```
Comment (sondy99) [inline Scripts/BaseScript/BaseScript.h]: CodeStyle: * close to the type... hehe check in general! (All files :) )

File: Source/ComponentTransform.cpp [hunk 4]
```diff
@@ -137,12 +136,24 @@ void ComponentTransform::SetGlobalTransform(const float4x4 & newglobal, const fl
 	UpdateOldTransform();
 }
 
+void ComponentTransform::SetPosition(const float3 & newPosition)
+{
+	position = newPosition;
+	gameobject->moved_flag = true;
```
Comment (sondy99) [inline Source/ComponentTransform.cpp]: moved_flag should be movedFlag

File: Source/GameObject.cpp [hunk 5]
```diff
@@ -138,15 +139,7 @@ void GameObject::Update()
 	}
 	for (std::list<GameObject*>::iterator it_child = children.begin(); it_child != children.end();)
```
Comment (sondy99) [inline Source/GameObject.cpp]: it_child should be itChild

File: Source/GameObject.cpp [hunk 6]
```diff
@@ -156,6 +149,16 @@ void GameObject::Update()
 			(*it_child)->UpdateBBox();
 			(*it_child)->moved_flag = false;
 		}
+
+		(*it_child)->Update(); //Update after moved_flag check
+
+		if ((*it_child)->copy_flag) //Copy GO
```
Comment (sondy99) [inline Source/GameObject.cpp]: change copy_flag to copyFlag.. in general use cameCase please :)

File: Source/ModuleScript.cpp [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+void ModuleScript::LoadFromMemory(int resource) //TODO: Load from memory in shipping build
+{
+	HRSRC hResource = FindResourceA(nullptr, MAKEINTRESOURCEA(resource), "DLL");
+	HGLOBAL hMemory = LoadResource(nullptr, hResource);
+
+	assert(hMemory != nullptr);
+	int size_bytes = SizeofResource(nullptr, hResource);
+	void *ptr = LockResource(hMemory);
```
Comment (sondy99) [inline Source/ModuleScript.cpp]: '*' close to the type :)

**Pull request merged**